### ISSUE TITLE
Foresees more than it should for Symfony Serializer

### ISSUE THREAD

**Issue by peter-gribanov** [phpstan/phpstan-strict-rules#114]
After installing this plugin, i found errors like:

```
 ------ ----------------------------------------------------------------------------------------------------------------------------------------
  Line   VideoServerFileInfoResolver.php
 ------ ----------------------------------------------------------------------------------------------------------------------------------------
  103    Instanceof between VideoFileInfo and VideoFileInfo will always evaluate to true.
 ------ ----------------------------------------------------------------------------------------------------------------------------------------
```

Snippet of code:

```php
use Symfony\Component\Serializer\SerializerInterface;

class VideoServerFileInfoResolver
{
    private SerializerInterface $serializer;

    public function __construct(SerializerInterface $serializer)
    {
        $this->serializer = $serializer;
    }

    public function getVideoFileInfo(): VideoFileInfo
    {
        // many code ...
        $data = '';
        // end of many code

        $file_info = $this->serializer->deserialize($data, VideoFileInfo::class, 'json');

        if (!$file_info instanceof VideoFileInfo) { // Error at this line !!!
            throw new \RuntimeException(sprintf('Failed deserialize video file info "%s".', $data));
        }

        return $file_info;
    }
}
```

PHPStan with this plugin suggests removing the check of the deserialization result, despite the fact that [the result can be an `object` or an `array`](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Serializer/SerializerInterface.php#L37), and not a specific class (`VideoFileInfo`).

We cannot disable this check. We cannot trust deserializers. Disabling this check will result in PHPStorm and [PHPStan](https://phpstan.org/r/44a03709-59ec-47e0-9b62-e431ee9b2601) errors.

![image](https://user-images.githubusercontent.com/1954436/94457092-cb3a3100-01bc-11eb-8349-c5dad0cbb262.png)

**Comment by ondrejmirtes** [comment]
Hi, the point of this extension is that it knows `$this->serializer->deserialize($data, VideoFileInfo::class, 'json')` will only ever return `VideoFileInfo`. It's smarter than PhpStorm in this regard. Do you have a situation when it's not true, e.g. how can you cause the `RuntimeException` to be thrown in real-life situation?

**Comment by peter-gribanov** [comment]
No. We have not encountered situations when another object is returned. Are you suggesting to ignore the error in the PHPStorm?

**Comment by hrach** [comment]
How exactly is the serializer error propagated? By returning a null? Then PHPStan should be fixed. By throwing? Then your code should try/catch.  In any case, [PhpStorm meta info](https://www.jetbrains.com/help/phpstorm/ide-advanced-metadata.html) may help storm to understand that it is returning a type passed as second argument.

**Comment by peter-gribanov** [comment]
> How exactly is the serializer error propagated?

It can throw [`NotEncodableValueException`](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Serializer/Serializer.php#L139) or [`NotNormalizableValueException`](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Serializer/Serializer.php#L208) if deserialization is not supported. If there is a configured denormalizer for these parameters, then the denormalization is delegated to it. The denomolizer can throw one of [these exceptions](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Serializer/Normalizer/DenormalizerInterface.php#L39-L45).
That is, the actual check of compliance of the deserialization result with the expected one is not performed by the Serializer and must be performed manually on the client.

**Comment by ondrejmirtes** [comment]
So it will throw an exception which isn't handled in your code, but it won't ever return a wrong value, right?

**Comment by peter-gribanov** [comment]
> In any case, PhpStorm meta info may help storm to understand that it is returning a type passed as second argument.

Thanks. This solved the problem for me. Maybe someone will need the code:

```php
# .phpstorm.meta.php
namespace PHPSTORM_META {
    // PHPStorm metadata directives
    override(\Symfony\Component\Serializer\SerializerInterface::deserialize(1), type(1));
    override(\Symfony\Component\Serializer\Normalizer\DenormalizerInterface::denormalize(1), type(1));
}
```

> So it will throw an exception which isn't handled in your code, but it won't ever return a wrong value, right?

If a specific deserializer for specific parameters is written correctly, then yes. The serializer always returns the correct value or throws an exception.

I think i can close this problem.

**Comment by amcsi** [comment]
A note on the above from @peter-gribanov:

If you grab the normalizer/serializer directly from the service container via service location (`$container->get(DenormalizerInterface::class)`) rather than dependency injection, PhpStorm is "not confident enough" that the returned value really is DenormalizerInterface or "mixed" so it does not apply the custom PhpStorm meta overrides.

In this case you need to make PhpStorm sure that the value is the normalizer by doing:

```php
// The docblock below is necessary.
/** @var DenormalizerInterface $normalizer */
$denormalizer = $container->get(DenormalizerInterface::class);
// $instance is inferred to be an instance of MyClassInstance.
$instance = $denormalizer->denormalize($array, MyClassInstance::class);
```
